File: src/main/java/com/flyfish233/gzip64/java/Gzip64.java
package com.flyfish233.gzip64.java;

import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import java.io.IOException;
import java.util.Base64;
import java.util.zip.GZIPInputStream;
import java.util.zip.GZIPOutputStream;

// @TargetApi(Build.VERSION_CODES.FROYO)
public class Gzip64 {
    public static String encode(String str) throws IOException {
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();
        GZIPOutputStream gzipOutputStream = new GZIPOutputStream(byteArrayOutputStream);
        gzipOutputStream.write(str.getBytes());
        gzipOutputStream.close();
        // Android
        // return Base64.encodeToString(byteArrayOutputStream.toByteArray(), Base64.DEFAULT);
        return Base64.getEncoder().encodeToString(byteArrayOutputStream.toByteArray());
    }

    public static String decode(String encodedStr) throws IOException {
        // Android
        // byte[] compressed = Base64.decode(encodedStr, Base64.DEFAULT);
        byte[] compressed = Base64.getDecoder().decode(encodedStr);
        ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(compressed);
        GZIPInputStream gzipInputStream = new GZIPInputStream(byteArrayInputStream);
        ByteArrayOutputStream byteArrayOutputStream = new ByteArrayOutputStream();

        byte[] buffer = new byte[1024];
        int len;
        while ((len = gzipInputStream.read(buffer)) != -1) {
            byteArrayOutputStream.write(buffer, 0, len);
        }
        byteArrayInputStream.close();

        return byteArrayOutputStream.toString();
    }
}

File: src/main/java/com/flyfish233/gzip64/kotlin/Gzip64.kt
package com.flyfish233.gzip64.kotlin

import okio.*
import kotlin.io.encoding.Base64
import kotlin.io.encoding.ExperimentalEncodingApi

object Gzip64 {
    fun encode(byteArray: ByteArray): String {
        val buffer = Buffer().write(byteArray)

        GzipSink(buffer).buffer().apply {
            write(buffer, buffer.size)
            close()
        }

        return buffer.readByteString().base64().also {
            buffer.close()
        }
    }

    fun encode(str: String): String = encode(str.toByteArray())

    @OptIn(ExperimentalEncodingApi::class)
    fun decodeToSrc(base64String: String): Source =
        GzipSource(Buffer().write(Base64.decode(base64String)))

    fun decodeToString(base64String: String): String =
        decodeToSrc(base64String).buffer().readUtf8()
}